File: Data Structures & Algorithms/subsets-ii/submission-5.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        
        res = []
        subset = []

        def solve(i):
            res.append(subset[::])

            for j in range(i, len(nums)):
                if j>i and nums[j] == nums[j-1]:
                    continue

                subset.append(nums[j])
                solve(j+1)
                subset.pop()

        nums.sort()
        solve(0)
        return res

File: Data Structures & Algorithms/three-integer-sum/submission-5.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        int n = nums.length;
        if(n == 0){
            return res;
        }

        Arrays.sort(nums);
        for(int i=0; i<n; i++){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            int l = i+1, r = n-1;
            while(l < r){
                if(l > i+1 && nums[l] == nums[l-1]){
                    l++;
                    continue;
                } else if(r < n-1 && nums[r] == nums[r+1]){
                    r--;
                    continue;
                }

                int sum = nums[i] + nums[l] + nums[r];
                if(sum == 0){
                    res.add(Arrays.asList(nums[i], nums[l], nums[r]));
                    l++; r--;
                } else if(sum < 0){
                    l++;
                } else {
                    r--;
                }
            }
        }

        return res;
    }
}
